### DISCUSSION TITLE
Go Extending framework: tutorial or fully working code

### DISCUSSION THREAD

**Discussion by nokia6290** [pocketbase/pocketbase#4247] [Q&A] [answered]
Hello,

Been trying to extend pocketbase as GO framework and it's a bigger challenge than I thought.. I'm not a GO, nor a JS developer and really lost in how the Go and pocketbase should be run, not to mention modified with additional custom routes. I spent about 8h already on this alone with no results, cant even run the project as its constantly plaguing with red errors of Go/Pocketbase.

By chance anyone could share their main.go file and extension with customRoutes?

I just want a starting prototype, and to add one single custom Route and then its done, but even this small feature on Pocketbase is amazingly difficult.

```go
package main

import (
	"log"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis_custom"
	"github.com/pocketbase/pocketbase/core"
)

func main() {
	app := pocketbase.New()

	// Add your custom route handlers
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.POST("/api/custom/workout", func(c echo.Context) error {
			return apis_custom.HandleInsertWorkout(app, c)
		})
		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

```go
package apis_custom

import (
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/models"
)

// HandleInsertWorkout handles the workout data insertion.
func HandleInsertWorkout(app *pocketbase.PocketBase, c echo.Context) error {
	var workoutData map[string]interface{}
	if err := c.Bind(&workoutData); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": err.Error()})
	}

	workoutRecord := models.NewRecord(app.Dao().Collection("workout"))
	workoutRecord.Load(workoutData)

	if err := app.Dao().SaveRecord(workoutRecord); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, workoutRecord)
}
```

My files, and they are constantly flagging red, either import, or god knows what else is wrong.

f.e. if I run go build
@MBP pocketbase2 % go build
apis_custom/custom_route_workout.go:7:2: import "github.com/pocketbase/pocketbase" is a program, not an importable package
found packages main (main.go) and pocketbase (pocketbase.go) in ....
@MBP pocketbase2 %

**Comment by ganigeorgiev** [comment, marked as answer]
Your `apis_custom` package file doesn't seem correct (ex. there is no such method `app.Dao().Collection("workout")` and also it is not recommend to persist untrusted user data without validations). Here is a slightly refactored version:

```go
// apis_custom/handler.go
package apis_custom

// imports...

func HandleInsertWorkout(app *pocketbase.PocketBase, c echo.Context) error {
	collection, err := app.Dao().FindCollectionByNameOrId("workout")
	if err != nil {
		return err
	}

	record := models.NewRecord(collection)

	// see https://pocketbase.io/docs/go-records/#create-new-record-with-data-validations
	form := forms.NewRecordUpsert(app, record)

	if err := form.LoadRequest(c.Request(), ""); err != nil {
		return apis.NewBadRequestError("Failed to read request data.", err)
	}

	if err := form.Submit(); err != nil {
		return apis.NewBadRequestError("Failed to create workout record.", err)
	}

	return c.JSON(http.StatusOK, record)
}
```


> is a program, not an importable package found packages main (main.go) and pocketbase (pocketbase.go) in ....

What is the name of your module in `go.mod`? Make sure that it doesn't conflict with other packages.

> but even this small feature on Pocketbase is amazingly difficult

If you find this _"amazingly difficult"_ you are bound to be disappointment, especially considering the lack of documentation and some not well thought out APIs at the moment.

In the end, you don't have to use PocketBase if you don't feel comfortable using Go.
There are plenty of Open Source alternatives out there that you can give a try, just to mention some:

- [Supabase](https://supabase.com/)
- [Appwrite](https://appwrite.io/)
- [Nhost](https://nhost.io/)
- [EdgeDB](https://www.edgedb.com/)
- [SurrealDB](https://surrealdb.com/)
- [PayloadCMS](https://payloadcms.com/)
- [Strapi](https://strapi.io/)
- [Directus](https://directus.io/)
- etc.

If you still want to extend PocketBase, I recommend going through [A Tour of Go](https://go.dev/tour/welcome/1).

**Reply by nokia6290** [reply]
Thanks, I will try your suggestions and explore other options.
